src/cls/rgw_gc/cls_rgw_gc.cc

```diff
@@ -114,13 +114,31 @@ static int cls_rgw_gc_queue_list_entries(cls_method_context_t hctx, bufferlist *
   }
 
   cls_rgw_gc_urgent_data urgent_data;
-  if (head.bl_urgent_data.length() > 0) {
-    auto iter_urgent_data = head.bl_urgent_data.cbegin();
-    try {
-      decode(urgent_data, iter_urgent_data);
-    } catch (ceph::buffer::error& err) {
-      CLS_LOG(5, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode urgent data\n");
-      return -EINVAL;
+  std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+  if (g_ceph_context->_conf->rgw_gc_defer_enabled) {
+    if (head.bl_urgent_data.length() > 0) {
+      auto iter_urgent_data = head.bl_urgent_data.cbegin();
+      try {
+        decode(urgent_data, iter_urgent_data);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(5, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode urgent data\n");
+        return -EINVAL;
+      }
+    }
+    if (urgent_data.num_xattr_urgent_entries > 0) {
+      bufferlist bl;
+      ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl);
+      if (ret < 0) {
+        CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+        return ret;
+      }
+      auto iter = bl.cbegin();
+      try {
+        decode(xattr_urgent_data_map, iter);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(1, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode xattrs urgent data map\n");
+        return -EINVAL;
+      }
```

Comment:
I feel that variations of this code are repeated a number of times. It would be nice if "urgent data" were abstracted into its own class, so the implementation details don't bleed everywhere and result in duplicated code. I realize it's a big ask, so it's only a suggestion, possibly for a future PR.